### ISSUE TITLE
Feature request: numpy arrays of objects as inputs to dask.compute

### ISSUE THREAD

**Issue by hmaarrfk** [dask/dask#3692]
Currently, I can create (nested) lists of objects that are a mix of eagerly computed items and delayed items.

If I pass that list to dask.compute, it can create the graph and computes the result as a new list replacing the delayed items with their computed counterparts.

The list has a very well defined structure that I would like to exploit. As such, before using Dask, I had been using numpy array with `dtype=object`.

The workaround, suggested by @mrocklin, is something like:
```
tasks = np.asarray(dask.compute(*tasks.tolist()))
```

Are there any plans to support numpy arrays as inputs to the `compute` function?
I understand it probably isn't an easy answer due to the fact that you should support nesting and whatnot.

[Code example](https://gist.github.com/hmaarrfk/21c6df997f9011f81930cd24ca549ed9)

[Stack overflow question](https://stackoverflow.com/questions/51109461/using-dask-compute-with-an-array-of-delayed-items) answered by @mrocklin Thanks!

**Comment by jakirkham** [comment]
Not sure if this makes sense for your use case, but there are a number of ways to [join Dask Arrays together into a single Dask Array]( dask.pydata.org/en/latest/array-stack.html ). It's also possible to [convert Delayed objects to Dask Arrays]( http://dask.pydata.org/en/latest/array-api.html#dask.array.from_delayed ). NumPy Arrays can also be converted, but that conversion often happens under the hood. Also [asarray]( http://dask.pydata.org/en/latest/array-api.html#dask.array.asarray ) can be used to coerce various array-like things including Python lists to Dask Arrays. Would be interested to hear if this helps or not.

**Comment by hmaarrfk** [comment]
@mrocklin, see PR that is referenced above. I check if the numpy array is a numpy array of objects first. In which case, it is basically equivalent to a dictionary (which may be arbitrary large as well).

**Comment by mrocklin** [comment]
Right, I think that the same logic holds.  It's much more likely that that
numpy array of objects holds a bunch of text.  Does this approach
significantly slow down that case?  My guess is that it does.

**Comment by hmaarrfk** [comment]
@mrocklin are you talking about [`numpy.chararray`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.chararray.html) or simply `np.empty((3, 3), dtype=object)` that happens to contain strings?

Maybe instead of checking for numpy arrays in the PR I submitted, I can check for dask arrays of objects?

**Comment by mrocklin** [comment]
Object arrays

import numpy as np
np.array(['a', 'bb', 'ccc'] * 1000000, dtype=object)

Or, more broadly, we need to make sure that features like this don't affect
performance in other common case workloads.  The example above is one
common case workload that I can think of that I suspect will be harmed.
There may also be others.

**Comment by hmaarrfk** [comment]
@jakirkham I haven't forgotten about you. I was trying to create a nice self contained example of what I thought your suggestion was. Here is the gist, the array isn't actually computed through the delayed interface: https://gist.github.com/hmaarrfk/3852d40bc82362499f48083092efb702  

@mrocklin totally true, that is not a use case I have come across myself.
Maybe the answer is to learn how to use the collections api and create my own array-like collection.

**Comment by mrocklin** [comment]
> that is not a use case I have come across myself

It's pretty common whenever people deal with text.  For example this data might be names of people in a Pandas series or numpy array.

**Comment by hmaarrfk** [comment]
@mrocklin got it

Seems like I should be able to create a dask collection that does what I want.
I'm trying to go over the documentation again, now that I understand a little more how to use dask.
